File: backend/app/db/models/post.py
from sqlalchemy import Column, Integer, String, Text, ForeignKey, DateTime, Boolean
from sqlalchemy.orm import relationship
from datetime import datetime

from .base import Base
from .post_tag import post_tags

class Post(Base):
    __tablename__ = "posts"
    
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(200), index=True)
    content = Column(Text)
    author_id = Column(Integer, ForeignKey("users.id"))
    section_id = Column(Integer, ForeignKey("sections.id"))
    category_id = Column(Integer, ForeignKey("categories.id"))
    is_hidden = Column(Boolean, default=False)  # 是否隐藏，默认为False
    created_at = Column(DateTime, default=datetime.now)
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now)
    is_deleted = Column(Boolean, default=False)  # 添加软删除标记
    deleted_at = Column(DateTime, nullable=True)  # 记录删除时间
    vote_count = Column(Integer, default=0)  # 添加点赞计数字段
    
    author = relationship("User", back_populates="posts")
    section = relationship("Section", back_populates="posts")
    category = relationship("Category", back_populates="posts")
    comments = relationship("Comment", back_populates="post", cascade="all, delete-orphan")
    tags = relationship("Tag", secondary=post_tags, back_populates="posts")
    votes = relationship("PostVote", back_populates="post")
    favorited_by = relationship("PostFavorite", back_populates="post")  # 添加被收藏关系 

File: backend/app/db/models/Post.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

// 帖子模式
const PostSchema = new Schema({
  title: {
    type: String,
    required: [true, '标题不能为空'],
    trim: true,
    maxlength: [200, '标题最多200个字符']
  },
  content: {
    type: String,
    required: [true, '内容不能为空'],
  },
  author: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  category: {
    type: Schema.Types.ObjectId,
    ref: 'Category',
    default: null
  },
  tags: [{
    type: Schema.Types.ObjectId,
    ref: 'Tag'
  }],
  view_count: {
    type: Number,
    default: 0
  },
  vote_count: {
    type: Number,
    default: 0
  },
  is_hidden: {
    type: Boolean,
    default: false
  },
  is_deleted: {
    type: Boolean,
    default: false
  },
  is_sticky: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: {
    createdAt: 'created_at', 
    updatedAt: 'updated_at'
  },
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// 虚拟属性：评论数
PostSchema.virtual('comments', {
  ref: 'Comment',
  localField: '_id',
  foreignField: 'post',
  justOne: false,
  count: false
});

// 索引
PostSchema.index({ title: 'text', content: 'text' });
PostSchema.index({ author: 1 });
PostSchema.index({ category: 1 });
PostSchema.index({ created_at: -1 });

// 添加view_count的方法
PostSchema.methods.incrementViewCount = async function() {
  this.view_count += 1;
  return this.save();
};

// 静态方法：获取热门帖子
PostSchema.statics.getHotPosts = function(limit = 5) {
  return this.find({ is_deleted: false, is_hidden: false })
    .sort({ view_count: -1 })
    .limit(limit)
    .populate('author', 'username avatar')
    .populate('category', 'name color');
};

// 查询钩子：自动填充作者和分类
PostSchema.pre('find', function() {
  this.populate('author', 'username avatar')
      .populate('category', 'name color')
      .populate('tags', 'name');
});

PostSchema.pre('findOne', function() {
  this.populate('author', 'username avatar')
      .populate('category', 'name color')
      .populate('tags', 'name');
});

module.exports = mongoose.model('Post', PostSchema); 